gobblin-data-management/src/main/java/gobblin/data/management/util/PathUtils.java

```diff
@@ -53,4 +57,16 @@ public static Path getPathWithoutSchemeAndAuthority(Path path) {
   public static Path withoutLeadingSeparator(Path path) {
     return new Path(StringUtils.removeStart(path.toString(), Path.SEPARATOR));
   }
+
+  /**
+   * Finds the deepest ancestor of input that is not a glob.
+   */
+  public static Path deepestNonGlobPath(Path input) {
+    Path commonRoot = input;
+
+    while(GLOB_TOKENS.matcher(commonRoot.toString()).find()) {
```

Comment:
For the first comment, no, we need to match against entire string. For example `/foo/*/bar` is a glob, even though the name doesn't include a glob. I will fix the second.